x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java

```diff
@@ -0,0 +1,308 @@
+            }
+            settings = settingsBuilder.build();
+            newEnv = new Environment(settings, env.configFile());
+        } else {
+            newEnv = env;
+            settings = env.settings();
+        }
+
+        ensureFileRealmEnabled(settings);
+        try {
+            final Hasher hasher = Hasher.resolve(XPackSettings.PASSWORD_HASHING_ALGORITHM.get(settings));
+            password = new SecureString(generatePassword(PASSWORD_LENGTH));
+            username = generateUsername();
+
+            final Path passwordFile = FileUserPasswdStore.resolveFile(newEnv);
+            final Path rolesFile = FileUserRolesStore.resolveFile(newEnv);
+            FileAttributesChecker attributesChecker = new FileAttributesChecker(passwordFile, rolesFile);
+
+            Map<String, char[]> users = FileUserPasswdStore.parseFile(passwordFile, null, settings);
+            if (users == null) {
+                throw new IllegalStateException("File realm configuration file [" + passwordFile + "] is missing");
+            }
+            users = new HashMap<>(users);
+            users.put(username, hasher.hash(password));
+            FileUserPasswdStore.writeFile(users, passwordFile);
+
+            Map<String, String[]> userRoles = FileUserRolesStore.parseFile(rolesFile, null);
+            if (userRoles == null) {
+                throw new IllegalStateException("File realm configuration file [" + rolesFile + "] is missing");
+            }
+            userRoles = new HashMap<>(userRoles);
+            userRoles.put(username, ROLES);
+            FileUserRolesStore.writeFile(userRoles, rolesFile);
+
+            attributesChecker.check(terminal);
+            final boolean forceExecution = options.has(force);
+            checkClusterHealthWithRetries(newEnv, terminal, 5, forceExecution);
+            executeCommand(terminal, options, newEnv);
+        } catch (Exception e) {
+            int exitCode;
+            if (e instanceof UserException) {
+                exitCode = ((UserException) e).exitCode;
+            } else {
+                exitCode = ExitCodes.DATA_ERROR;
+            }
+            throw new UserException(exitCode, e.getMessage());
+        } finally {
+            cleanup(terminal, newEnv);
+        }
+    }
+
+    /**
+     * Removes temporary file realm user from users and roles file
+     */
+    private void cleanup(Terminal terminal, Environment env) throws Exception {
+        final Path passwordFile = FileUserPasswdStore.resolveFile(env);
+        final Path rolesFile = FileUserRolesStore.resolveFile(env);
+        FileAttributesChecker attributesChecker = new FileAttributesChecker(passwordFile, rolesFile);
+
+        Map<String, char[]> users = FileUserPasswdStore.parseFile(passwordFile, null, env.settings());
+        if (users == null) {
+            throw new UserException(ExitCodes.CONFIG, "File realm configuration file [" + passwordFile + "] is missing");
```

Comment:
Should we try to clean UserRolesStore anyway?

Reply:
Yes good idea. I originally thought that worst case scenario is that the roles files remain but reference a non-existent user which is ok. It is a bit sloppy and it's a similar comment to what Albert made on the original PR about trying to make these 2 changes atomic. I will rework this